### DISCUSSION TITLE
How to prevent auto closing the current tab?

### DISCUSSION THREAD

**Discussion by Dean-Corso** [uBlockOrigin/uAssets#24776] [Q&A]
Hi guys,

I would like to ask a question about an problem I have for a long time and I can not find a solution for that. So as the title already said, I have a problem with somehow randomly closing current working tab when clicking on any link. I really don't know what the problem is for that behavior but it happens few times a day and sometimes etc like a random. Even it happens when surfing on a forum and just clicking on next web page / next button to load the next site. Sometimes it happens that this tab then just closed right away after the click. Somehow this is really annoying. 

Just imagine I surf on forum and read some article and when trying to go forward to next page the entire tab is gone and I have to reopen it again. Mostly in those cases I'm using the middle mouse button to load the next page in a new tab to prevent that problem but this isn't really a good solution. So I would like to know whether this problem behavior is already known and whether there was already found a solution for that or not? I even don't know whether problem is uBO itself or anything you know. I just don't get a rid of that issue. I'm using latest official uBO version 1.59.0 in Brave browser. Maybe anyone of you could help to find & fix the problem. Thank you.

**Comment by stephenhawk8054** [comment]
Give the exact URL where you see the issue.

**Comment by krystian3w** [comment]
To me, it looks like a browser/system infection with something, to the lesser extent that one of the forums is hollow and can optionally be cured with network filters or scriptlets.

**Comment by Dean-Corso** [comment]
Hi guys, 

as I said, the problem just happens somehow randomly anywhere so there is no specific website. Problem is I can't force that issue anyhow to analyze it and even when I have open the uBO logger I can't see anything because the current tab is gone and all logging data of this current tab are also gone when the tab get closed.

So my question was generally and trying to find ways / methods to find the possible reason for that issue. I'm also not sure whether the problem is uBO itself or anything else (browser settings / extensions / TM scripts etc). So do you guys have any idea how to catch the problem anyhow? Is there any kind of logger about the action what happens in browser I could enable and let run in background and when the tab closing problem happens to get the reason logged? Just wanna find out what is closing my current tab sometimes / randomly. Maybe any exception gets not handled correctly because of any unknown reason and browser tab just get closed / terminated or whatever. As I said, that problem is pretty annoying when it happens. Any ideas what I could try & test are welcome guys.

PS: I'm using Brave browser which also has some shields to block ADS / FP etc (not using any ADS scripts or own rules). I have this enabled but without scripts. Maybe there is any issue in combination with uBO. Would you recommend to disable it completely? I did that now & removed two extensions which was flagged by Goog because of any violation (Fingerprint extension & MultiLogin). Will see whether I get still that problem or not (will send new results later). Thank you.

**Reply by stephenhawk8054** [reply]
Turn off uBO and test again.

**Reply by garry-ut99** [reply]
> Dean-Corso: I even don't know whether problem is uBO itself or anything you know
> Dean-Corso: I'm also not sure whether the problem is uBO itself or anything else (browser settings / extensions / TM scripts etc). 

This is not a general computer help desk forum, we solve problems related to uBO or caused by uBO. Before reporting anything here, turn off uBO and check whether the issue still occurs or not, don't report issues unrelated to uBO, you can report them in plenty of other general computer forums.

**Reply by Dean-Corso** [reply]
So I did turn off uBO yesterday and the problem didn't occur anymore. Not sure whether it was just a random but otherwise it really seems that uBO is problem and does close the tab. Maybe I should test it some more days without using uBO to be more sure that uBO is possibly the problem but it looks so.

Yes it's clear, so I wouldn't report the problem if I wouldn't think that uBO is the problem (first thought). So lets suppose that uBO is problem so how to find out what uBO does when the tab get closed? How to debug that issue? How to log or something? So the command to close the tab must come from any source like extension or browser etc. So what would you do in that case to find it?

**Reply by garry-ut99** [reply]
Have you followed **all** the steps from the template? https://github.com/uBlockOrigin/uAssets/issues/new?assignees=&labels=&projects=&template=bug_report.yml

Also paste your troubleshooting information, like explained in https://github.com/uBlockOrigin/uAssets/issues/20586#issue-1987866378 .

**Reply by Dean-Corso** [reply]
So the problem still didn't occur yet anymore since I have disabled uBO, I will test it few more days over the weekend to be more sure. Problem could be any enabled filter list or my custom filter list I made myself over the years. I can enable / disable them optionally to see whether the problem disappears but also then I don't know what possible rule/s did trigger the problem. Normally just using the popup / popunder syntax does close tab/windows so maybe there is an issue that even current tabs getting closed sometimes.

So what I can say is that I'm using uBO without any other ADS blocker similar extension except that Brave browser internal shields but there I don't use any script or custom rules (all blank) etc. When using uBO then I have enabled some specific scripts from filter lists and also using my custom filters too.

The main problem that I can't find out what the reason is why the tab get closed. How to log that? Would be nice when uBO had an optional internal logger we could enable to log specific actions. Otherwise I don't know how to find the reason for that behavior you know.

**Reply by stephenhawk8054** [reply]
Reset uBO to default settings. uBO has the logger: https://github.com/gorhill/uBlock/wiki/The-logger but you need to open it before reproducing the issue.

**Reply by garry-ut99** [reply]
> Dean-Corso : https://github.com/uBlockOrigin/uAssets/discussions/24776#discussioncomment-10268572 : Problem is I can't force that issue anyhow to analyze it and even when I have open the uBO logger I can't see anything because the current tab is gone and all logging data of this current tab are also gone when the tab get closed.

Did you have logger opened before the issue has occured or you opened logger afterwards? If the latter case, then try to open logger before the issue occurs, and in case you had it already open, but the log has still disappeared despite it, then try to click Page selector https://github.com/gorhill/uBlock/wiki/The-logger#page-selector and select again the tab that has been closed (it sometimes happens they are still here as per https://github.com/uBlockOrigin/uBlock-issues/issues/755 or https://github.com/uBlockOrigin/uBlock-issues/issues/2473 edit: I see you use Brave (Chromium), so it might be not applicable to your case, but maybe try on Firefox) or select "Active tab" or "All", maybe with a bit of luck there is a possibility it will bring the logs back.

> Dean-Corso : I can enable / disable them optionally to see whether the problem disappears but also then I don't know what possible rule/s did trigger the problem. 

Exactly, this is a standard procedure, which every user should do before reporting any issue: to try to find out a filter list which might cause the issue, we can then help you to track specific filters responsible for the issue.

**Reply by gorhill** [reply]
> what the reason is why the tab get closed

The only way uBO will close a tab is because of a `popup` filter or because of the no-popups per-site switch.

**Reply by garry-ut99** [reply]
> Dean-Corso: Title : How to prevent auto closing the current tab?
> Dean-Corso: https://github.com/uBlockOrigin/uAssets/discussions/24776#discussion-7016104 : I have a problem with somehow randomly closing current working tab when clicking on any link

As `popup` filters or `no-popups per-site` switch, close only new tab/windows, and not the current tab, it indicates that OP's issue is not caused by uBlock popup filters.

**Reply by gorhill** [reply]
> randomly closing current working tab when clicking on any link

`popunder` filters could cause this.

**Reply by Dean-Corso** [reply]
Some update guys. Yesterday I was checking out my custom filters and it seems the problem is right in there and you are right @gorhill , the problem is the popunder command. 😃 I could force that tab closing problem when trying to enter anything in a search bar and after few times (randomly) the tab does just close itself after pressing enter. So the thing I don't understand is why this happens just randomly.

I was using that basic filter below...(removed all websites I did added for domain / denyallow).
`*^$script,inline-script,other,ping,popunder,3p,websocket,domain=website.com,denyallow=site.com`

Strange thing is that the websites are not included in "domain" list to apply the blocking filters like popunder. Somehow this popunder command seems not to work so good (just tested in Brave only yet / maybe it works better in Firefox) and I removed that command now to prevent those random current tab closing. Maybe you could investigate that popunder command whether it could has any issues etc. So without using popunder the problem seems to be gone and there is also no ADS / popup window / tab poping up anywhere happen.

About the uBO logger. Of course I did had open the logger and set the view to display all logging entry's of entire action and when the tab closing did happen uBO did just log the first entry = website?search=123 with "doc" type and that was all. I'm not the expert here but I really think there is any issue with that popunder command itself or in combination with browser or something. Otherwise I can't see a reason why the current tab should get closed just so I'm working with you know. If this action is made on purpose then I'm not a fan of it. If users are just working with a single tab and browser setting is set to default (last tab get closed = browser exit call) then browser says goodbye and the user don't know what happened etc you know. Maybe some other handling would be better instead to prevent that current tab closing behavior. 

Otherwise uBO is a great tool I'm using for long time to create styling / hiding rules. Just love it and I send you all thumbs up 👍 and say thank you to Raymond & Co.

EDIT: So I did notice that also without using the popunder command in my custom filter list anywhere it can happen that the current tab randomly get closed when I click on any own bookmark link to open this bookmark link in the current tab (left mouse click). Hhmm. 😞

**Reply by garry-ut99** [reply]
> EDIT:: So I did notice that also without using the popunder command in my custom filter list anywhere it can happen that the current tab randomly get closed when I click on any own bookmark link to open this bookmark link in the current tab (left mouse click). Hhmm

Disable "My Filters", and try again, to determine whether it's your filters or some other filters (from filter lists) causing it.
If the issue still happens, disable different filter lists one by one, until the issue stops.

**Reply by Dean-Corso** [reply]
That's not so easy @garry-ut99, but I can try it to see what happens in next days. Otherwise lets think that any of my custom filters is the problem (even when its not the popup / popunder I did disabled all) how to find the problem rule? I have many rules in my own filter list so I can't check every single one and wait whether the problem happens or not. In that case some other logger would be much better or just a logger what does log the tab closing action. Is there nothing what I could use? Or is it possible to block entire tab closing from outside (only allow it manually by mouse click on exit X button of tab) sending commands etc? Just don't know what I could do.

**Reply by garry-ut99** [reply]
> Dean-Corso: I have many rules in my own filter list so I can't check every single one and wait whether the problem happens or not

You have filters, not rules, in your filter list, better to use a proper nomenclature and don't mix terms:
filters are filters https://github.com/gorhill/uBlock/wiki/Dashboard:-My-filters, 
and rules are rules https://github.com/gorhill/uBlock/wiki/Dashboard:-My-rules, 
(like mentioned at the bottom of https://github.com/uBlockOrigin/uBlock-issues/discussions/3208#discussioncomment-9111912)

No need to disable every single filter, in such cases you can just use [bisection method](https://en.wikipedia.org/wiki/Bisection_method), in simple words: 
- comment out half of your filters (it doesn't matter whether the top or the bottom part)
   - if the issue has stopped occuring it means the problematic filter is inside this part, 
   - if the issue has not stopped occuring, it means the problematic filter is in the other part
      - every time you track the problematic part, comment out half of it and check again
      - repeat the process until you find the problematic filter

I use this method sometimes, last time I used it successfully in: https://github.com/uBlockOrigin/uBOL-home/issues/140#issuecomment-2184192718

Besides, some filters like `$all` https://github.com/gorhill/uBlock/wiki/Static-filter-syntax#all might still block `$popups` (but not `$popunders`), for example disabling all filter lists, and then with only this filter `$all,~doc,domain=vipbox1.com` in My Filters visiting `vipbox1.com` and clicking a banner on the right will still cause a new tab to be closed.  `$doc` or normal filters like: `||toptrendyinc.com^` have higher priority than `$popup` filters, so enabling EasyList will cause the popup to not be closed, but only blocked, unless you disable it with `||toptrendyinc.com^$badfilter`. Though I don't say you have `$all,~doc` or similiar filters in you filters.

> Dean-Corso: Or is it possible to block entire tab closing from outside (only allow it manually by mouse click on exit X button of tab) sending commands etc? 

> Dean-Corso: If users are just working with a single tab and browser setting is set to default (last tab get closed = browser exit call) then browser says goodbye and the user don't know what happened etc you know. 

For the latter, maybe try, (not sure whether it's useful for the former question):
- https://chromewebstore.google.com/detail/chrome-close-lock/plcabbfeeokakkmdecdccmibahigjkno
- https://chromewebstore.google.com/detail/dont-close-window-with-la/dlnpfhfhmkiebpnlllpehlmklgdggbhn

**Reply by Dean-Corso** [reply]
Hello @garry-ut99,

thank you for the answer and trying me to help to find my problem. So you are right, I'm just using custom filters.

So using the half method sound logically but could take a longer while because I'm using almost 2000 lines of custom filters (not all enabled) and they are somehow messy. I will check this out even if I still didn't found a 100% method to force that tab closing problem to repeat it with enabled / disabled filters etc. That random tab closing is really a PITA. Just don't know why the there is no simple way to hook that function which is used to close a tab to see whether its coming from (like a stack check in debugger etc or exception logger).

About the $all filter, so I'm using that filter too on specific domains so I will try to disable them too next to popup & popunder for testing phase. Thanks for the hint.

One more question I have about that blocking syntax "||" so when do you recommend using that "||" and when not? In some filters I don't use || like when creating a blocking filter using domain= / denlyallow= syntax like this...
`*$script,inline-script,other,ping,3p,websocket,domain=123.com,denyallow=abc.com`
...instead of this...
`||*$script,inline-script,other,ping,3p,websocket,domain=123.com,denyallow=abc.com`
....just because in case of using || above the || are showing disabled / ghost and not colored as it does when using a website / URL right after. Is there a different now using || in that case?

**Reply by stephenhawk8054** [reply]
You can read more about the syntax in the documentations here:

- https://github.com/gorhill/uBlock/wiki/Static-filter-syntax
- https://help.eyeo.com/en/adblockplus/how-to-write-filters
- https://kb.adguard.com/en/general/how-to-create-your-own-ad-filters

**Comment by Thaodan** [comment]
It's part of the Facebooks internal help. The link can be found in the
help or through an internet search.

I suspect that Facebook only reacts like that if opened from another
website like a search engine.

**Reply by stephenhawk8054** [reply]
I mean, how can I reproduce the issue? What are the exact steps to reproduce?

**Comment by Thaodan** [comment]
1. Search 'Facebook Help' in your favorite search engine (DuckDuck go in my
case)
2. Open the first link pointing to Facebook help
3. Watch as Facebook closes the current tab and opens the link in a new
one

**Reply by stephenhawk8054** [reply]
I can't reproduce. Video recording:

https://github.com/user-attachments/assets/499d418e-3a77-4546-af67-960f380c4fb0

**Comment by Thaodan** [comment]
Hm maybe I have not enabled some rules.

These are the rules I have enabled:
added:
  adguard-generic: 97243-4266, 2d.17h.10m
  adguard-mobile: 10811-59, 2d.17h.10m
  adguard-spyware-url: 2040-137, 2d.17h.10m
  ublock-annoyances: 5892-57, 2h.39m Δ
  dpollock-0: 11737-396, 15h.22m
  FIN-0: 2219-6, 1h.39m
  adguard-spyware: 156443-29790, 2d.17h.10m
 default:
  user-filters: 51-1, never
  ublock-filters: 40530-124, 2h.39m Δ
  ublock-badware: 11620-7, 2h.39m Δ
  ublock-privacy: 2771-21, 2h.39m Δ
  ublock-unbreak: 2623-1, 2h.39m Δ
  ublock-quick-fixes: 248-13, 2h.39m Δ
  easylist: 69204-1536, 2h.39m Δ
  easyprivacy: 53538-149, 2h.39m Δ
  urlhaus-1: 34380-0, 14h.22m
  plowe-0: 3451-1358, 5d.15h.39m
  DEU-0: 7951-52, 14h.22m

**Reply by stephenhawk8054** [reply]
Test again in a new browser profile with only uBO and its default settings. Don't sync anything.
